datafusion/functions/src/string/concat.rs

```diff
@@ -88,26 +307,82 @@ impl ScalarUDFImpl for ConcatFunc {
         &self.signature
     }
 
+    fn coerce_types(&self, arg_types: &[DataType]) -> Result<Vec<DataType>> {
+        use DataType::*;
+
+        if arg_types.is_empty() {
+            return plan_err!("concat requires at least one argument");
+        }
+
+        // Arrays need no coercion
+        for dt in arg_types {
+            if matches!(dt, List(_) | LargeList(_) | FixedSizeList(_, _)) {
+                return Ok(arg_types.to_vec());
+            }
+        }
```

Comment:
If we find one list type we exit early; what happens if we have a `List, String` input? Shouldn't we capture that error here?